application/common/models/Email.php

```diff
@@ -0,0 +1,284 @@
+<?php
+
+namespace common\models;
+
+use Exception;
+use yii\behaviors\TimestampBehavior;
+use yii\helpers\ArrayHelper;
+use yii\web\ServerErrorHttpException;
+
+/*
+ * Email is a Yii2 model for a database email queue
+ *
+ * Copied from silinternational/email-service
+ */
+class Email extends EmailBase
+{
+    /** int $delay number of seconds to delay sending */
+    public $delay_seconds = 0;
+
+    public function scenarios()
+    {
+        return [
+            self::SCENARIO_DEFAULT => [
+                'to_address',
+                'cc_address',
+                'bcc_address',
+                'subject',
+                'text_body',
+                'html_body',
+                'delay_seconds',
+                'send_after',
+            ],
+        ];
+    }
+
+    public function rules()
+    {
+        return ArrayHelper::merge(
+            parent::rules(),
+            [
+                [
+                    'attempts_count', 'default', 'value' => 0,
+                ],
+                [
+                    ['to_address', 'cc_address', 'bcc_address'],
+                    'email',
+                    'message' => '{attribute} is not a valid email address: {value}',
+                ],
+                [
+                    'text_body', 'required', 'when' => function ($model) {
+                        return empty($model->html_body);
+                    },
+                ],
+            ]
+        );
+    }
+
+    public function behaviors()
+    {
+        // http://www.yiiframework.com/doc-2.0/yii-behaviors-timestampbehavior.html
+        return [
+            TimestampBehavior::class,
+        ];
+    }
+
+    /**
+     * Attempt to send email. Returns 1 on success.
+     * @throws Exception if sending failed
+     * @return int number of sent messages
+     */
+    public function send(): int
+    {
+        /*
+         * Try to send email or throw exception
+         */
+        $message = $this->getMessage();
+        if (!$message->send()) {
+            throw new Exception('Unable to send email', 1461011826);
```

Comment:
I don't know it's worth making this change, but it might be beneficial to put a new timestamp in here, so that as we search our codebase for a given error code, only one line of code in a single repo will match. That could avoid confusion regarding which repo (email-service or idp-id-broker) is the source of the problem. There should be other ways to tell (such as which application's logs we see that error code in), but I just thought I'd mention it in case you also think it's worth updating.

Reply:
Same with the other error codes copied from email-service.